addon/mixins/adapter-fetch.ts

```diff
@@ -0,0 +1,187 @@
+  let headersObject: PlainObject = {};
+
+  if (headers) {
+    headers.forEach((value, key) => (headersObject[key] = value));
+  }
+
+  return headersObject;
+}
+
+export default Mixin.create({
+  headers: undefined as undefined | PlainHeaders,
+  /**
+   * @override
+   */
+  ajaxOptions(url: string, type: Method, options: object): FetchOptions {
+    let hash = (options || {}) as AjaxOptions;
+    hash.url = url;
+    hash.type = type;
+
+    // Add headers set on the Adapter
+    let adapterHeaders = get(this, 'headers');
+    if (adapterHeaders) {
+      hash.headers = assign(hash.headers || {}, adapterHeaders);
+    }
+
+    const mungedOptions = mungOptionsForFetch(hash);
+
+    // Mimics the default behavior in Ember Data's `ajaxOptions`, namely to set the
+    // 'Content-Type' header to application/json if it is not a GET request and it has a body.
+    if (
+      mungedOptions.method !== 'GET' &&
+      mungedOptions.body &&
+      (mungedOptions.headers === undefined ||
+        !(
+          mungedOptions.headers['Content-Type'] ||
+          mungedOptions.headers['content-type']
+        ))
+    ) {
+      mungedOptions.headers = mungedOptions.headers || {};
+      mungedOptions.headers['Content-Type'] = 'application/json; charset=utf-8';
+    }
+
+    return mungedOptions;
+  },
+
+  /**
+   * @override
+   */
+  ajax(url: string, type: Method, options: object) {
+    const requestData = {
+      url,
+      method: type
+    };
+
+    const hash = this.ajaxOptions(url, type, options);
+
+    return (
+      this._ajaxRequest(hash)
+        // @ts-ignore
+        .catch((error, response, requestData) => {
+          throw this.ajaxError(this, response, null, requestData, error);
+        })
+        .then((response: Response) => {
+          return RSVP.hash({
+            response,
+            payload: determineBodyPromise(response, requestData)
+          });
+        })
+        .then(
+          ({
+            response,
+            payload
+          }: {
+            response: Response;
+            payload: string | object | undefined;
+          }) => {
+            if (response.ok) {
+              return this.ajaxSuccess(this, response, payload, requestData);
+            } else {
+              throw this.ajaxError(this, response, payload, requestData);
+            }
+          }
+        )
+    );
+  },
+
+  /**
+   * Overrides the `_ajaxRequest` method to use `fetch` instead of jQuery.ajax
+   * @override
+   */
+  _ajaxRequest(
+    options: Mix<RequestInit, { url: string }>
+  ): RSVP.Promise<Response> {
+    return this._fetchRequest(options.url, options);
+  },
+
+  /**
+   * A hook into where `fetch` is called.
+   * Useful if you want to override this behavior, for example to multiplex requests.
+   */
+  _fetchRequest(url: string, options: RequestInit): RSVP.Promise<Response> {
+    return fetch(url, options);
+  },
+
+  /**
+   * @override
+   */
+  ajaxSuccess(
+    adapter: any,
```

Comment:
Can we do better than `any` here? maybe a `DS.Adapter`?

Reply:
> The best you could do is explicitly specify `this` in your method signatures.

We don't need this, `this` has a type here already
```js
static create<T, Base = EmberObject>(
    args?: MixinOrLiteral<T, Base> & ThisType<Fix<T & Base>>
): Mixin<T, Base>;
```

This is `Mixin.create` signature, so technically I can write `Mixin.create(FetchAdapter, DS.RESTAdapter>`

The issue is that

1. I have to manually define the object literal type
2. There might be **other** mixins, so type of `this` will still be inaccurate at runtime.

But I think it's still better than doing `adapter: any` here. Updated it to have an interface.